runtime/src/tokio.rs

```diff
@@ -471,22 +446,31 @@ impl crate::Sink for Sink {
 /// Implementation of [`crate::Stream`] for the `tokio` runtime.
 pub struct Stream {
     context: Context,
-    stream: SplitStream<Framed<TcpStream, LengthDelimitedCodec>>,
+    stream: OwnedReadHalf,
 }
 
 impl crate::Stream for Stream {
-    async fn recv(&mut self) -> Result<Bytes, Error> {
-        let result = timeout(self.context.executor.cfg.read_timeout, self.stream.next())
+    async fn recv(&mut self, buf: &mut [u8]) -> Result<usize, Error> {
+        // Wait for the stream to be readable
+        timeout(self.context.executor.cfg.read_timeout, self.stream.readable())
```

Comment:
This timeout should apply to the full message read? (that's what it was doing before)